Code/GraphMol/SynthonSpaceSearch/SynthonSpace.cpp

```diff
@@ -0,0 +1,719 @@
+    const auto &rxn = reaction.second;
+    size_t thisSize = 1;
+    for (size_t i = 0; i < rxn->getSynthons().size(); ++i) {
+      thisSize *= rxn->getSynthons()[i].size();
+    }
+    totSize += thisSize;
+  }
+  return totSize;
+}
+
+SubstructureResults SynthonSpace::substructureSearch(
+    const ROMol &query, SynthonSpaceSearchParams params) {
+  PRECONDITION(query.getNumAtoms() != 0, "Search query must contain atoms.");
+
+  if (params.randomSample) {
+    if (!d_randGen) {
+      std::random_device rd;
+      d_randGen.reset(new std::mt19937(rd()));
+    }
+    if (params.randomSeed != -1) {
+      d_randGen->seed(params.randomSeed);
+    }
+  }
+  std::vector<std::unique_ptr<ROMol>> results;
+  RDKit::MatchVectType dontCare;
+
+  auto fragments = details::splitMolecule(query, params.maxBondSplits);
+  std::vector<SynthonSpaceHitSet> allHits;
+  size_t totHits = 0;
+  for (auto &fragSet : fragments) {
+    auto theseHits = searchFragSet(fragSet);
+    if (!theseHits.empty()) {
+      totHits += std::accumulate(
+          theseHits.begin(), theseHits.end(), 0,
+          [&](const size_t prevVal, const SynthonSpaceHitSet &hs) -> size_t {
+            return prevVal + hs.numHits;
+          });
+      allHits.insert(allHits.end(), theseHits.begin(), theseHits.end());
+    }
+  }
+
+  if (params.buildHits) {
+    std::sort(allHits.begin(), allHits.end(),
+              [&](const SynthonSpaceHitSet &hs1,
+                  const SynthonSpaceHitSet &hs2) -> bool {
+                if (hs1.reactionId == hs2.reactionId) {
+                  return hs1.numHits < hs2.numHits;
+                } else {
+                  return hs1.reactionId < hs2.reactionId;
+                }
+              });
+    // Keep track of the result names so we can weed out duplicates by
+    // reaction and synthons.  Different splits may give rise to the same
+    // synthon combination.  This will keep the same molecule produced via
+    // different reactions which I think makes sense.  The resultsNames will
+    // be accumulated even if the molecule itself doesn't make it into the
+    // results set, for example if it isn't a random selection or it's
+    // outside maxHits or hitStart.
+    std::set<std::string> resultsNames;
+    // The random sampling, if requested, doesn't always produce the required
+    // number of hits in 1 sweep.  Also, totHits, being an upper bound, may
+    // not be achievable.
+    size_t tmpMaxHits(params.maxHits);
+    int numSweeps = 0;
+    while (results.size() < std::min(tmpMaxHits, totHits) && numSweeps < 10) {
+      buildHits(allHits, query, params, totHits, resultsNames, results);
+      // totHits is an upper bound, so may not be reached.
+      if (params.maxHits == -1 || !params.randomSample) {
+        break;
+      }
+      numSweeps++;
+    }
+  }
+  return SubstructureResults{std::move(results), totHits};
+}
+
+namespace {
+inline std::vector<std::string> splitLine(const std::string &str,
```

Comment:
Maybe use boost::tokenizer instead of rolling your own? There are a few examples of this elsewhere in the RDKit code

Reply:
It's just a concealed, simple, use of std::sregex_token_iterator.  I don't see any benefit in using boost for this.